python_modules/dagster/dagster/_core/definitions/freshness_based_auto_materialize.py

```diff
@@ -102,136 +101,104 @@ def get_execution_period_and_conditions_for_policies(
             if (
-                key not in target_asset_keys_and_parents
-                or key not in asset_graph.materializable_asset_keys
-                or not asset_graph.get_downstream_freshness_policies(asset_key=key)
+                isinstance(asset_graph, ExternalAssetGraph)
+                and AssetKeyPartitionKey(parent_key) in will_materialize_mapping[parent_key]
             ):
-                continue
-
-            parents = asset_graph.get_parents(key)
-
-            if any(p in waiting_to_materialize for p in parents):
-                # we can't materialize this asset yet, because we're waiting on a parent
-                waiting_to_materialize.add(key)
-                continue
-
-            # if we're going to materialize a parent of this asset that's in a different repository,
-            # then we need to wait
-            if isinstance(asset_graph, ExternalAssetGraph):
-                repo = asset_graph.get_repository_handle(key)
-                if any(
-                    AssetKeyPartitionKey(p, None) in conditions_by_asset_partition
-                    and asset_graph.get_repository_handle(p) is not repo
-                    for p in parents
-                ):
-                    waiting_to_materialize.add(key)
-                    continue
-
-            # figure out the current contents of this asset with respect to its constraints
-            current_data_time = data_time_resolver.get_current_data_time(key, current_time)
-
-            # figure out the expected data time of this asset if it were to be executed on this tick
-            # for root assets, this would just be the current time
-            expected_data_time = (
-                min(
-                    (
-                        cast(datetime.datetime, expected_data_time_by_key[k])
-                        for k in parents
-                        if k in expected_data_time_by_key
-                        and expected_data_time_by_key[k] is not None
-                    ),
-                    default=None,
-                )
-                if asset_graph.has_non_source_parents(key)
-                else current_time
+                parent_repo = asset_graph.get_repository_handle(parent_key)
+                if parent_repo != asset_graph.get_repository_handle(asset_key):
+                    return {}, set(), current_data_time
+            # find the minimum non-None data time of your parents
+            parent_expected_data_time = expected_data_time_mapping.get(
+                parent_key
+            ) or data_time_resolver.get_current_data_time(parent_key, current_time)
+            expected_data_time = min(
+                filter(None, [expected_data_time, parent_expected_data_time]),
+                default=None,
             )
+    else:
+        expected_data_time = current_time
+
+    # currently, freshness logic has no effect on partitioned assets
+    if not asset_graph.is_partitioned(asset_key):
```

Comment:
could we short circuit sooner if partitioned

Reply:
yep good call